airbyte-integrations/connectors/source-harvest/source_harvest/streams.py

```diff
@@ -0,0 +1,299 @@
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        params["per_page"] = self.per_page
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        stream_data = response.json()
+
+        # depending on stream type we may get either:
+        # * nested records iterable in response object;
+        # * not nested records iterable;
+        # * single object to yield.
+        if self.data_field:
+            stream_data = response.json().get(self.data_field, [])
+
+        if isinstance(stream_data, list):
+            yield from stream_data
+        else:
+            yield stream_data
+
+
+class IncrementalHarvestStream(HarvestStream, ABC):
+    cursor_field = "updated_at"
+
+    def __init__(self, updated_since: pendulum.datetime = None, **kwargs):
+        super().__init__(**kwargs)
+        self._updated_since = updated_since
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        updated_since = stream_state.get(self.cursor_field) or self._updated_since
+        params.update({"updated_since": updated_since})
+        return params
+
+
+class HarvestSubStream(HarvestStream):
+    @property
+    @abstractmethod
+    def path_template(self) -> str:
+        """
+        :return: sub stream path template
+        """
+
+    @property
+    @abstractmethod
+    def parent_stream(self) -> IncrementalHarvestStream:
+        """
+        :return: parent stream class
+        """
+
+    def stream_slices(self, sync_mode: SyncMode, cursor_field: List[str] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        items = self.parent_stream(authenticator=self.authenticator)
+        for item in items.read_records(sync_mode=sync_mode):
```

Comment:
is it intentional that sync_mode is the input `sync_mode` and not full_refresh? this way if we ever sync a parent entity then I think we don't pick up later updates to it because the sync mode is FR right?

Reply:
it's just a vestige of prev code version. I've changed that